pkg/server/status_insights.go

```diff
@@ -0,0 +1,178 @@
+)
+
+// StmtInsightsDetails represents the structure of details column
+// in system.statement_execution_insights table.
+// Reminder for future maintainers: keep this struct backward compatible.
+// Do not remove/rename already existing fields.
+type StmtInsightsDetails struct {
+	RowsRead    int64 `json:"rows_read"`
+	RowsWritten int64 `json:"rows_written"`
+}
+
+func (d *StmtInsightsDetails) ToJSON() (string, error) {
+	b, err := json.Marshal(d)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+// StatementExecutionInsights requests statement insights that satisfy specified
+// parameters in request payload if any provided.
+func (s *statusServer) StatementExecutionInsights(
+	ctx context.Context, req *serverpb.StatementExecutionInsightsRequest,
+) (*serverpb.StatementExecutionInsightsResponse, error) {
+	ctx = authserver.ForwardSQLIdentityThroughRPCCalls(ctx)
+	ctx = s.AnnotateCtx(ctx)
+	if err := s.privilegeChecker.RequireViewActivityOrViewActivityRedactedPermission(ctx); err != nil {
+		return nil, err
+	}
+	// Validate that either both start and end time are defined or not.
+	if (req.StartTime != nil && req.EndTime == nil) || (req.EndTime != nil && req.StartTime == nil) {
+		return nil, errors.New("required that both StartTime and EndTime to be set")
+	}
+
+	resp := &serverpb.StatementExecutionInsightsResponse{
+		StatementInsights: make([]*serverpb.StatementExecutionInsightsResponse_StatementInsight, 0),
+	}
+
+	// Request in-memory statement insights in case persisted data is incomplete.
+	inMemoryInsights, err := s.ListExecutionInsights(ctx, &serverpb.ListExecutionInsightsRequest{})
+	if err != nil {
+		return nil, err
+	}
+	contentionEventsResp, err := s.TransactionContentionEvents(ctx, &serverpb.TransactionContentionEventsRequest{})
+	if err != nil {
+		return nil, err
+	}
+	stmtIdToContentionMap := make(map[clusterunique.ID][]contentionpb.ExtendedContentionEvent)
+	for _, event := range contentionEventsResp.Events {
+		// users with VIEWACTIVITYREDACTED role have no access to range keys so
+		// it will be empty.
+		if event.BlockingEvent.Key.Equal(roachpb.Key{}) {
+			continue
+		}
+		stmtIdToContentionMap[event.WaitingStmtID] = append(stmtIdToContentionMap[event.WaitingStmtID], event)
+	}
+	for _, insight := range inMemoryInsights.Insights {
+		for _, stmt := range insight.Statements {
+			if req.StatementID != nil && stmt.ID != *req.StatementID {
+				continue
+			}
+			if req.StmtFingerprintID != appstatspb.StmtFingerprintID(0) && stmt.FingerprintID != req.StmtFingerprintID {
+				continue
+			}
+			if req.StartTime != nil && req.EndTime != nil &&
+				!timeutil.IsOverlappingTimeRanges(*req.StartTime, *req.EndTime, stmt.StartTime, stmt.EndTime) {
+				continue
+			}
+
+			if events, ok := stmtIdToContentionMap[stmt.ID]; ok && stmt.Contention != nil {
+				for _, event := range events {
+					ci, err := s.getContentionEventDetails(&event)
+					if err != nil {
+						return nil, err
```

Comment:
Not sure if we want to fail the entire request if we are unable to get contention details. Maybe we can fail more gracefully and provide partial responses.